File: m3meka/python/m3/dynamatics.py
#M3 -- Meka Robotics Robot Components
#Copyright (c) 2010 Meka Robotics
#Author: edsinger@mekabot.com (Aaron Edsinger)

#M3 is free software: you can redistribute it and/or modify
#it under the terms of the GNU Lesser General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#M3 is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU Lesser General Public License for more details.

#You should have received a copy of the GNU Lesser General Public License
#along with M3.  If not, see <http://www.gnu.org/licenses/>.

import yaml
import os 
from m3.toolbox import *
import m3.dynamatics_pb2 as mrt
import m3.toolbox as m3t
import m3.toolbox_ctrl as m3tc
from m3.component import M3Component
from m3.unit_conversion import *
import numpy as nu
from PyKDL import *
import scipy.linalg

class M3Dynamatics(M3Component):
    """Interface for joint-space controllers of a M3Sea
    """
    def __init__(self,name,type='m3dynamatics'):
        M3Component.__init__(self,name,type=type)
        self.status=mrt.M3DynamaticsStatus()
        self.command=mrt.M3DynamaticsCommand()
        self.param=mrt.M3DynamaticsParam()
        
	for i in range(3):
	    self.param.payload_com.append(0)
	for i in range(6):
            self.param.payload_inertia.append(0)
	    
	self.read_config()	

	self.T80=nu.zeros([4,4],nu.float32) #Transform from end to base frame
	self.T08=nu.zeros([4,4],nu.float32)  #Transform from base to end frame
	self.G=nu.zeros(self.ndof+1,nu.float32) #Gravity vector on joints
	self.C=nu.zeros(self.ndof+1,nu.float32) #Coriolis vector on joints
	self.J=nu.zeros([6,self.ndof],nu.float32) #Jacobian from joint torques to end torques
	self.Jt=nu.zeros([self.ndof,6],nu.float32)#Transform end wrench to joint torques.
	self.end_twist=nu.zeros(6,nu.float32)
	self.end_pos=nu.zeros(3,nu.float32)
	self.end_rot=nu.zeros([3,3],nu.float32)
	#self.set_tool_transform(nu.identity(4,nu.float32)) #default
	
		
    #Utility API
    def get_end_twist(self): 
	return self.end_twist
   
	
    # ##########################################################
    """ End and Eff frame related functions 
    Assumes that wrench is 6x1 Numeric array. 
    Assumes that positions are [x,y,z,1] Numeric array. 
    All units in meters/mNm/mN/radians/seconds unless otherwise noted.
    """
    def get_end_force(self):
	print 'get_end_force not yet implemented'
	return nu.zeros(3,nu.float32)
	#return self.end_wrench[:3]
	
    def get_end_moment(self):
	print 'get_end_moment not yet implemented'
	return nu.zeros(3,nu.float32)
	#return self.end_wrench[3:]
	
    def get_end_velocity(self):
	return self.end_twist[:3]
    
    def get_end_angular_velocity(self):
	return self.end_twist[3:]
    
    def get_end_roll_pitch_yaw_rad(self):
	return nu.array(self.__get_end_rotation_kdl().GetRPY(),float)
    
    def get_end_euler_zyx_rad(self):
	return nu.array(self.__get_end_rotation_kdl().GetEulerZYX(),float)
    
    def get_end_euler_zyx_deg(self):
	return rad2deg(self.get_end_euler_zyx_rad())
    
    def get_end_euler_zyz_rad(self):
	return nu.array(self.__get_end_rotation_kdl().GetEulerZYZ())
    
    def get_end_euler_zyz_deg(self):
	return rad2deg(self.get_end_euler_zyz_rad())
        
    def __get_end_rotation_kdl(self):
	c = self.end_rot
	return Rotation(c[0,0],c[0,1],c[0,2],c[1,0],c[1,1],c[1,2],c[2,0],c[2,1],c[2,2])
    
    def get_end_roll_pitch_yaw_deg(self):
	return rad2deg(self.get_end_roll_pitch_yaw_rad())
    
    def get_end_position(self):
	return self.end_pos
    
    def get_end_position_In(self):
	return m2in(self.end_pos)
    
    def end_wrench_2_joint_torques(self,w):
	return nu.dot(self.Jt*nu.array(w))
    
    def joint_torques_2_end_wrench(self,tq):
	#return nu.dot(self.J,nu.array(tq))
	print 'get_end_force not yet implemented'
	return nu.zeros(6,nu.float32)
    
    def thetadot_2_end_twist(self,td):
	return nu.dot(self.J,nu.array(td))
    
    def eff_frame_2_base_frame(self,x):
	return nu.dot(self.T80,x)
    
    def base_frame_2_eff_frame(self,x):
	return nu.dot(self.T08,x)
      
	    
    
    # ##########################################################
    """ Tool frame related functions 
     Assumes that wrench is 6x1 Numeric array. 
     Assumes that positions are [x,y,z,1] Numeric array. 
     All units in meters/mNm/radians/seconds unless otherwise noted.."""
    
    def set_tool_transform(self,T):
	""" Set the homgenous transform from tool frame to end frame"""
	self.T2E = nu.array(T,nu.float32) #Transform point in hand frame to arm tool.
	self.E2T =  nu.transpose(self.T2E)  #Transform point in arm tool to hand frame.
	self.FE2FT = m3tc.force_moment_transform(self.E2T)  #Transform wrench in end frame to tool frame
	self.FT2FE = m3tc.force_moment_transform(self.T2E)  #Transform wrench in tool frame to end frame
	
    def tool_wrench_2_joint_torques(self,wrench):
	return self.end_wrench_2_joint_torques(self.tool_wrench_2_end_wrench(wrench))
    
    def joint_torques_2_tool_wrench(self,tq):
	return self.end_wrench_2_tool_wrench(self.joint_torques_2_end_wrench(tq))
    
    def tool_wrench_2_end_wrench(self,wrench):
	return nu.matrixmultiply(self.tool_wrench_2_end_wrench_transform(),wrench)
    
    def end_wrench_2_tool_wrench(self,wrench):
	return nu.matrixmultiply(self.end_wrench_2_tool_wrench_transform(),wrench)
    
    def end_wrench_2_eff_wrench_transform(self):
	T=self.T08.copy()
	T[0:3,3]=0
	return m3tc.force_moment_transform(T)
    
    def eff_wrench_2_end_wrench_transform(self):
	T=self.T80.copy()
	T[0:3,3]=0
	return m3tc.force_moment_transform(T)
    
    def tool_wrench_2_end_wrench_transform(self):
	return nu.matrixmultiply(self.eff_wrench_2_end_wrench_transform(),self.FT2FE)
    
    def end_wrench_2_tool_wrench_transform(self):
	return nu.matrixmultiply(self.FE2FT,self.end_wrench_2_eff_wrench_transform())
	
	
    def base_frame_2_tool_frame(self,x):
	return nu.dot(self.E2T,self.base_frame_2_eff_frame(x))
    
    def tool_frame_2_base_frame(self,x):
	return self.eff_frame_2_base_frame(nu.dot(self.T2E,x))
    # ##########################################################
    
    def get_torque_gravity_mNm(self): 
	return self.G
    def get_torque_gravity_inLb(self): 
	return mNm2inLb(self.G)
    
    def set_payload_com(self,com): #meters
	"""Set the payload center-of-mass (tool Frame)"""
	self.param.payload_com[0]=float(com[0])
	self.param.payload_com[1]=float(com[1])
	self.param.payload_com[2]=float(com[2])
    def set_payload_mass(self,m): #Kg
	"""Set the estimated payload mass"""
	self.param.payload_mass=m
 
    def get_end_rotation(self):	
	return self.end_rot
    
    def get_end_frame(self):
	return self.T80
	
    def update_status(self):
	
	self.end_rot = self.list_to_numpy_rotation(self.status.end_rot)	
	self.end_pos = nu.array(self.status.end_pos,nu.float32)	
	self.T80[:3,:3] = self.end_rot
	self.T80[0,3] = self.end_pos[0]
	self.T80[1,3] = self.end_pos[1]
	self.T80[2,3] = self.end_pos[2]	
	self.T80[3,3] = 1	
	self.T08=scipy.linalg.inv(self.T80)
	self.G=nu.array(self.status.G,nu.float32)	
	self.J=nu.array(self.status.J,nu.float32).resize([6,self.ndof])
	self.Jt=nu.transpose(self.J)	
	self.end_twist=nu.array(self.status.end_twist,nu.float32)
	
    def list_to_numpy_rotation(self, l):
	mtx = mtx = nu.zeros([3,3])
	mtx[0,0] = l[0]
	mtx[0,1] = l[1]
	mtx[0,2] = l[2]
	mtx[1,0] = l[3]
	mtx[1,1] = l[4]
	mtx[1,2] = l[5]
	mtx[2,0] = l[6]
	mtx[2,1] = l[7]
	mtx[2,2] = l[8]
	return mtx

    def read_config(self):
	M3Component.read_config(self)
	
	try:
            f=file(self.config_name,'r')
            config= yaml.safe_load(f.read())
        except (IOError, EOFError):
            print 'Config file not present:',self.config_name
            return
	
	self.ndof = config['ndof']
        
    

File: m3meka/src/m3/chains/dynamatics.cpp
/* 
M3 -- Meka Robotics Robot Components
Copyright (c) 2010 Meka Robotics
Author: edsinger@mekabot.com (Aaron Edsinger)

M3 is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

M3 is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with M3.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "m3/chains/dynamatics.h"


namespace m3
{
	
using namespace m3rt;
using namespace std;
using namespace KDL;

bool M3Dynamatics::LinkDependentComponents()
{	
	m3chain = (M3JointChain*)factory->GetComponent(chain_name);
	
	if (m3chain!=NULL)
		return true;
	else
	{
		M3_ERR("M3Dynamatics: no chain found with name: %s\n", chain_name.c_str());
	}
	return false;
}

void M3Dynamatics::Startup()
{		
	if (ndof != m3chain->GetNumDof())
	{
		M3_WARN("M3Dynamatics: ndof does not match chain: %s\n", chain_name.c_str());		
	}
	
	q.resize(ndof+3);	// contains both position q.q and vel q.qdot
	qdot_id.resize(ndof+3);
	qdotdot_id.resize(ndof+3);		
	G.resize(ndof+3);
	J=Jacobian(ndof+3);	
	chain_mass=0;
	for (int i=0; i<ndof; i++)
	{
		status.add_g(0);
		chain_mass += m[i];
	}

	for (int i=0; i<3; i++)
	{
		status.add_end_pos(0);
	}

	for (int i=0; i<9; i++)
	{
		status.add_end_rot(0);
	}

	for (int i=0; i<6; i++)
	{
		status.add_end_twist(0);
		status.add_base_wrench(0);
		status.add_child_wrench(0);
	}

	for (int i=0; i<(6*ndof); i++)
	{
		status.add_j(0);
	}
	
	for (int i=0; i<=ndof; i++)
	{		
		// KDL adds each joint to the BASE of each frame, not the end as Roboop did
		Frame frame = Frame().DH_Craig1989(a[i], DEG2RAD(alpha[i]), d[i], DEG2RAD(joint_offset[i]));
		
		// The first segment is static defining the offset from the origin,
		// but we are adding a joint in each direction that will always be zero to calc
		// wrenches at the base
		
		if (i==0) 
		{
			kdlchain.addSegment(Segment(Joint(Joint::RotX)));
			kdlchain.addSegment(Segment(Joint(Joint::RotY)));
			kdlchain.addSegment(Segment(Joint(Joint::RotZ), frame));
		} else {			
			// because frames define the joint placement for the NEXT segment, we need to use Inertial values for previous segment
			Vector vcog = Vector(cx[i-1] ,cy[i-1], cz[i-1]);					
			RotationalInertia ri = RotationalInertia(Ixx[i-1], Iyy[i-1], Izz[i-1], Ixy[i-1], Ixz[i-1], Iyz[i-1]);
			kdlchain.addSegment(Segment(Joint(Joint::RotZ), frame, frame.Inverse()*RigidBodyInertia(m[i-1],vcog, ri)));
		}
	}
	
	//Store no-load payload for last link
	Frame toTip = kdlchain.getSegment(kdlchain.getNrOfSegments()-1).getFrameToTip();
	RigidBodyInertia z_I_rigid = toTip*kdlchain.getSegment(kdlchain.getNrOfSegments()-1).getInertia();
	z_com = z_I_rigid.getCOG();
	z_I = z_I_rigid.getRotationalInertia();
	z_m = z_I_rigid.getMass();
	
	fksolver_vel = new ChainFkSolverVel_recursive(kdlchain);
	grav = Vector(0,0,1.)*GRAV;
	idsolver = new ChainIdSolver_RNE(kdlchain, grav);
	f_ext = std::vector<Wrench>(kdlchain.getNrOfSegments());
	jjsolver = new ChainJntToJacSolver(kdlchain); 

	SetPayload();
	SetStateSafeOp();
}



void M3Dynamatics::Shutdown()
{

}

void M3Dynamatics::StepCommand()
{
	
}

void M3Dynamatics::StepStatus()
{	
	if (IsStateError())
		return;
	// ToDo: find out why copy operator= causes lockups in hard RT	
	for (int i=0; i<3; i++)
	{
		q.q(i) = 0.;
		q.qdot(i) = 0.;
		qdotdot_id(i) = 0.;
		qdot_id(i) = 0.;
	}	
	for (int i=0; i<ndof; i++)
	{
		q.q(i+3) = m3chain->GetThetaRad(i);
		q.qdot(i+3) = m3chain->GetThetaDotRad(i);
		if (param.use_accelerations())
		{
			qdotdot_id(i+3) = m3chain->GetThetaDotDotRad(i);
		}
		else
			qdotdot_id(i+3) = 0.;
		if (param.use_velocities())
			qdot_id(i+3) = m3chain->GetThetaDotRad(i);
		else
			qdot_id(i+3) = 0.;

	}	
	SetPayload();	
	idsolver->SetGrav(grav);
	
	f_ext[kdlchain.getNrOfSegments()-1] = end_wrench;	
	
	int result = idsolver->CartToJnt(q.q, qdot_id, qdotdot_id, f_ext, G);
	
	for (int i=0; i<G.rows(); i++)
		G(i) = -1000*G(i);  // Nm to mNm
	if (result==-1)
		M3_ERR("ID solver returned error %d for M3Kinestatics component %s\n", result, GetName().c_str());
	jjsolver->JntToJac(q.q, J);
	fksolver_vel->JntToCart(q, end_2_base_framevel);
	//fksolver_pos->JntToCart(q.q,T80);
	end_twist = Twist(end_2_base_framevel.p.v, end_2_base_framevel.M.w);
	end_pos = end_2_base_framevel.p.p;
	end_rot = end_2_base_framevel.M.R;
	base_wrench.force = grav*GetMass();
	base_wrench.torque = Vector(G(0),G(1),G(2))/1000.;
	
	for (int i=0; i<ndof; i++){
		status.set_g(i, G(i+3));
	}
	
	for (int i=0; i<3; i++)
		status.set_end_pos(i, end_pos(i));

	for (int i=0; i<6; i++)
	{
		if (i<3)
		{
			status.set_end_twist(i, end_twist.vel(i));
			status.set_base_wrench(i, base_wrench.force(i));
			status.set_child_wrench(i, end_wrench.force(i));
		} else {
			status.set_end_twist(i, RAD2DEG(end_twist.rot(i-3)));
			status.set_base_wrench(i, base_wrench.torque(i-3));
			status.set_child_wrench(i, end_wrench.torque(i-3));
		}
	}
	for (int i=0; i<9; i++)
		status.set_end_rot(i, end_rot.data[i]);

	for (int i=0; i<6; i++)
	{
		for (int j=0; j<ndof; j++)
			status.set_j((i*ndof)+(j), J(i,j+3));
	}

	for (int i=0; i<ndof; i++)
		m3chain->SetG(i, GetG(i));
	m3chain->SetGCoupled();	
}

// This should be realtime safe... Explanation:
// When first created in StartUp idsolver creates a vector of segments for the chain allocating necessary memory.
// When the constructor is called again it resizes the vector to zero, however when it rebuilds it it should
// not exceed the size of the previous allocation and not reallocate any memory.
void M3Dynamatics::SetPayload()
{
	if (param.payload_inertia_size()==6)
	{		
		//param.payload_inertia defined as Ixx, Ixy, Ixz, Iyy, Iyz, Izz
		// KDL defines RotationalInertia(Ixx,Iyy,Izz,Ixy,Ixz,Iyz);
		payload_I = RotationalInertia(param.payload_inertia(0),
						param.payload_inertia(3),
						param.payload_inertia(5),
						param.payload_inertia(1),
						param.payload_inertia(2),
						param.payload_inertia(4));
	}
	else
	{
		//Possible bug, size changes from 0-3. Likely need of mutex. Appears to be fixed. Keep this just in case.
		M3_ERR("Bug: Bad field size in M3Dynamatics::payload_inertia...%d instead of  6\n", param.payload_inertia_size());
		PrettyPrint();
	}
	if (param.payload_com_size()==3)
	{
		for (int i; i<3; i++)
			payload_com(i)=param.payload_com(i);		
	}
	else
	{
		//Possible bug, size changes from 0-3. Likely need of mutex. Appears to be fixed. Keep this just in case.
		M3_ERR("Bug: Bad field size in M3Dynamatics::payload_com...%d instead of 3 \n", param.payload_com_size());
		PrettyPrint();
	}

	Frame toTip = kdlchain.getSegment(kdlchain.getNrOfSegments()-1).getFrameToTip();
	Segment * end_eff = kdlchain.getMutableSegment(kdlchain.getNrOfSegments()-1);
	mReal m = GetPayloadMass();
	Vector com = toTip*GetPayloadCom(); // tranforming from wrist to last joint's frame where we defined it's COM...
	RotationalInertia rot_inertia = (toTip*RigidBodyInertia(0.,Vector(0.,0.,0.),GetPayloadInertia())).getRotationalInertia();
						
	if (m+z_m>0.001)
	{
		Vector ecom = (m*com+z_com*z_m)/(m+z_m);		
		end_eff->setInertia(toTip.Inverse()*RigidBodyInertia(m+z_m, ecom, rot_inertia + z_I));
	}
	else
	{
		Vector ecom = (com+z_com);		
		end_eff->setInertia(toTip.Inverse()*RigidBodyInertia(m+z_m, ecom, rot_inertia + z_I));
	}
		
	idsolver->chain = kdlchain;
}
///////////////////////////////////////////////////////////////////////////////////////////////////////////////////

bool M3Dynamatics::ReadConfig(const char * filename)
{
	if (!M3Component::ReadConfig(filename))
		return false;
	
	YAML::Node doc;
	GetYamlDoc(filename, doc);
	
	mReal rtemp;
	int itemp;
	
	doc["chain_component"] >> chain_name;	
	doc["ndof"] >> ndof;	
	
	const YAML::Node& ymlparam = doc["param"];

	ymlparam["payload_mass"] >> rtemp;
	param.set_payload_mass(rtemp);	
	for(int i=0; i<ymlparam["payload_com"].size(); i++)
	{
		ymlparam["payload_com"][i] >> rtemp;
		param.add_payload_com(rtemp);
		payload_com(i)=rtemp;
	}

	for(int i=0; i<ymlparam["payload_inertia"].size(); i++)
	{
		ymlparam["payload_inertia"][i] >> rtemp;
		param.add_payload_inertia(rtemp);
	}

	payload_I = RotationalInertia(param.payload_inertia(0),
						param.payload_inertia(3),
						param.payload_inertia(5),
						param.payload_inertia(1),
						param.payload_inertia(2),
						param.payload_inertia(4));

	bool btemp;
	try
	{
	    ymlparam["use_velocities"] >> btemp;
	    param.set_use_velocities(btemp);
	} catch(YAML::TypedKeyNotFound<string> e) 
	{
		param.set_use_velocities(0);
	} 
	try
	{
	  ymlparam["use_accelerations"] >> btemp;
	  param.set_use_accelerations(btemp);
	} catch(YAML::TypedKeyNotFound<string> e) 
	{
		param.set_use_accelerations(0);
	} 	


	const YAML::Node& links = doc["links"];

	for(int i=0; i<links.size(); i++)
	{		
		links[i]["d"] >> rtemp;
		d.push_back(rtemp);
		links[i]["a"] >> rtemp;
		a.push_back(rtemp);
		links[i]["alpha"] >> rtemp;
		alpha.push_back(rtemp);
		links[i]["joint_offset"] >> rtemp;
		joint_offset.push_back(rtemp);
		if (i < (links.size()-1))
		{
			links[i]["m"] >> rtemp;
			m.push_back(rtemp);
			links[i]["cx"] >> rtemp;
			cx.push_back(rtemp);
			links[i]["cy"] >> rtemp;
			cy.push_back(rtemp);
			links[i]["cz"] >> rtemp;
			cz.push_back(rtemp);
			if(const YAML::Node *pName = links[i].FindValue("Ixx")) {
    				*pName >> rtemp;
			} else {
    				rtemp = 0.0;
			}
			Ixx.push_back(rtemp);
			if(const YAML::Node *pName = links[i].FindValue("Ixy")) {
    				*pName >> rtemp;
			} else {
    				rtemp = 0.0;
			}
			Ixy.push_back(rtemp);
			if(const YAML::Node *pName = links[i].FindValue("Ixz")) {
    				*pName >> rtemp;
			} else {
    				rtemp = 0.0;
			}
			Ixz.push_back(rtemp);
			if(const YAML::Node *pName = links[i].FindValue("Iyy")) {
    				*pName >> rtemp;
			} else {
    				rtemp = 0.0;
			}
			Iyy.push_back(rtemp);
			if(const YAML::Node *pName = links[i].FindValue("Iyz")) {
    				*pName >> rtemp;
			} else {
    				rtemp = 0.0;
			}
			Iyz.push_back(rtemp);
			if(const YAML::Node *pName = links[i].FindValue("Izz")) {
    				*pName >> rtemp;
			} else {
    				rtemp = 0.0;
			}
			Izz.push_back(rtemp);
		}
	}

	return true;
}

}